### ISSUE TITLE
PKCS#12 support in OpenSSL 3.0 FIPS provider

### ISSUE THREAD

**Issue by vikas-yadav77** [openssl/openssl#19997]
We are using pkcs#12 container to store personal identification information and when I am trying to use OpenSSL-3.0.0 in FIPS mode to load PKCS#12 file the following error is coming

root@mackinac-6ull:/data/Openssl_LS_test_cert# openssl pkcs12 -in new.pfx -info
Enter Import Password:
MAC: sha256, Iteration 2048
MAC length: 32, salt length: 8
**Error verifying PKCS12 MAC; no PKCS12KDF support.**
Use -nomacver if MAC verification is not required.
root@mackinac-6ull:/data/Openssl_LS_test_cert#

Is PKCS#12 not FIPS supported? And if it is not FIPS compliant then is there any other FIPS certified container that we can use to store personal identification information.
Is there any way we can use FIPS approved algorithms in PKCS#12?

**Comment by vikas-yadav77** [comment]
@mattcaswell  Thanks for quick response. 
So it means that we can not use PKCS#12 container in FIPS mode. Is this understanding correct?

is PKCS12KDF algorithm is hardcoded in PCKS#12 standard or we can change the KDF fnuncion at time of .pfx creation?

**Comment by beldmit** [comment]
There is a draft 
https://datatracker.ietf.org/doc/draft-ietf-lamps-pkcs12-pbmac1/ by @tomato42 and I have a very draft implementation of it.

**Comment by mattcaswell** [comment]
> So it means that we can not use PKCS#12 container in FIPS mode. Is this understanding correct?

Correct. At least not without one of the workarounds I mentioned

> is PKCS12KDF algorithm is hardcoded in PCKS#12 standard or we can change the KDF fnuncion at time of .pfx creation?

It's hard coded (but see @beldmit's post above).

**Comment by vikas-yadav77** [comment]
@mattcaswell @beldmit Thanks for the updates.

@mattcaswell, If the draft gets approved and PKCS#12 comes with a new implementation, does OpenSSL require any changes in the FIPS provider code to integrate the new changes (PKCS#12)?

I am asking this question because if there is any change in FIPS provider code then OpenSSL have to go to CMWP again for new FIPS 140-2 certification.

**Comment by mattcaswell** [comment]
> @mattcaswell, If the draft gets approved and PKCS#12 comes with a new implementation, does OpenSSL require any changes in the FIPS provider code to integrate the new changes (PKCS#12)?

It's unclear. The proposed draft is based on PBMAC1 which combines a password based KDF (PBKDF2) with an underlying MAC (and at a minimum a compliant implementation must support SHA-256 HMAC). Our current FIPS module already supports PBKDF2 and SHA-256 HMAC. However the FIPS module (and OpenSSL in general) does not have a MAC implementation known as PBMAC1. The code that combines PBKDF2 with SHA-256 HMAC is actually quite trivial, but I'm not sure whether that code would need to be inside the module boundary or not. If not then the current module already has everything needed. If it does then modifications to the module (and hence a revalidation would be required).

In any case this is likely to be a long term solution. The draft would have to be approved, OpenSSL would have to implement it, and it would have to be incorporated into an OpenSSL release (OpenSSL 3.3 at the earliest I would think) before it could be used.

**Comment by tomato42** [comment]
I'm pretty sure that "PBMAC1" doesn't have to be in the fips module, otherwise all constructions that use some kind of KDF to get keys for the HMAC would have to be inside the fips module (e.g. the TLS ciphers).

**Comment by mattcaswell** [comment]
> I'm pretty sure that "PBMAC1" doesn't have to be in the fips module, otherwise all constructions that use some kind of KDF to get keys for the HMAC would have to be inside the fips module (e.g. the TLS ciphers).

Good point. Hopefully that's true. I guess we'd need to confirm it with the lab.

**Comment by paulidale** [comment]
If we implement it as a new MAC, it would need to be inside the boundary or a work-around like @mattcaswell mentioned would be required.

**Comment by t8m** [comment]
> If we implement it as a new MAC

Not sure why we would want to do it. It is clear that naming it as a "special" MAC is just for the purpose of adding an object identifier. We can special-case it in the code.

**Comment by beldmit** [comment]
The only reason I see is that the intermediate data shouldn't cross the FIPS boundary.

**Comment by WilburZjh** [comment]
I am looking at this page of docs -> https://www.openssl.org/docs/man3.0/man1/openssl-pkcs12.html. Its mentioned in the `nomac` option that 

`Do not attempt to provide the MAC integrity. This can be useful with the FIPS provider as the PKCS12 MAC requires PKCS12KDF which is not an approved FIPS algorithm and cannot be supported by the FIPS provider.`

This makes me confused, is PKCS12 MAC not supported by FIPS or PKCS12KDF not supported by FIPS?

**Comment by WilburZjh** [comment]
My openssl version is OpenSSL 1.1.1k  FIPS 25 Mar 2021. May I know what is the difference if I create a pkcs12 keystore with and without the nomac option? It seems that I can create two keystores successfully within a fips version openssl. Does that mean the mac value is calculated in a fips version openssl?

**Comment by t8m** [comment]
The OpenSSL 1.1.1k  FIPS 25 Mar 2021 is a vendor (most probably Red Hat) patched version, not something that the OpenSSL project directly provides. We cannot really give you any guidance for this FIPS module. 

The `-nomac` option creates a PKCS12 file without an integrity protection. The PKCS12KDF is used to create the key for the MAC. Unfortunately that KDF function is not approved by FIPS standard which means, if strict compliance with FIPS rules is required, that it cannot be used. That's why our FIPS module does not provide that KDF function and thus it cannot create PKCS12 files with integrity protection.

**Comment by WilburZjh** [comment]
Thanks for the information. General speaking, does PKCS12 need a MAC check in FIPS mode?

**Comment by t8m** [comment]
> General speaking, does PKCS12 need a MAC check in FIPS mode?

Without doing the MAC check, you're losing the integrity protection on the PKCS12 file. It depends on how do you use it. If you need just confidentiality and not the integrity protection, you do not need to do MAC check. You can also provide the integrity protection on transport by other means such as adding a signature outside of the PKCS12 file.

**Comment by tomato42** [comment]
> Thanks for the information. General speaking, does PKCS12 need a MAC check in FIPS mode?

yes, FIPS rules require use of integrity protection with confidentiality protection for storage of key material; by omitting integrity protection you need to consider the keys to be stored in plantext

**Comment by raghupcr** [comment]
@mattcaswell @paulidale 
Hi Matt/Paul,

PKCS12KDF was working when we were using FIPS 1.0.2 + FIPS and now we migrated to 3.0.8 + FIPS, now it is not able to load the certificates in urlib3. So is there a change from the previous version of openssl w.r.t FIPS algorithms?

**Comment by paulidale** [comment]
That KDF cannot be FIPS validated so it isn't in the FIPS provider.

You can still load such by using the default provider and an explicit 'fips=no' property query.

**Comment by raghupcr** [comment]
> That KDF cannot be FIPS validated so it isn't in the FIPS provider.
> 
> You can still load such by using the default provider and an explicit 'fips=no' property query.

Thanks @paulidale 

PKCS12KDF is not working in OpenSSL 3.0.8 + FIPS ( As we are loading only fips + base provider, no default provider and inorder to use that we need to use a different context). This is fine

But PKCSKDF is working in OpenSSL 1.0.2 + FIPS this is our doubt, this is a bit concern here


so is there any difference here?

**Comment by mattcaswell** [comment]
>So is there a change from the previous version of openssl w.r.t FIPS algorithms?

1.0.2 does not really have the concept of KDF crypto algorithms as a separate thing, i.e. there are no EVP_KDF APIs there. The fact that the PKCS12 KDF works with the FIPS module there is probably "by accident". The 3.0 FIPS provider is a complete rewrite and works totally differently to the old FIPS module. In 3.0 the PKCS12 KDF is recognised as a crypto algorithm in its own right and does exist inside the FIPS provider (only in the default provider).

> can we load both default and fips provider in the default context in openssl 3.0.8 + fips use case? and make the pkcs12kdf work?

You could, and it would then work (dependent on property settings). Having both the default and fips providers loaded means you don't know which provider is used for any given algorithm since all algorithms in the fips provider are also in default. You can resolve this with appropriate global property settings. But ultimately loading some crypto from "default" means you are no longer FIPS.

**Comment by raghupcr** [comment]
> 

@mattcaswell 

Thanks for the information.

one doubt on this
You can resolve this with appropriate global property settings. But ultimately loading some crypto from "default" means you are no longer FIPS.
- we are loading as of now fips and base providers and making the fips enable call from the code as shown below
- self.crypto_context.EVP_default_properties_enable_fips(None, 1):

we understood that it is no longer fips compliant just for understanding a bit more, if we are using both providers when we are using the pkcs12kdf will it go to fips provider/default provider?

**Comment by raghupcr** [comment]
@paulidale @mattcaswell 

Hi Matt/Dr.Paul,

one of our counterparts in our organization where we call them to provide some certs (pub/priv/passphrase) information which we store and later use for connecting them (https) when we are on the customer site. This is where we are facing issues now.

They are using the below Java code to generate the certificate pair and pass it to our application.

snippet:
/ generate PKCS10 certificate request
        X500Principal subjectName = new X500Principal(cn);
        PKCS10CertificationRequestBuilder p10Builder = new JcaPKCS10CertificationRequestBuilder(subjectName, keyPair.getPublic());

so the way certificate generation using pkcs10 is the issue and when we use the contents in the urllib method of python which internally uses (https along with certs ) is getting failed in FIPS mode and working in non-FIPS mode

 self.__connection = PoolManager(cert_file=cert_file, key_file=cert_key, key_password=key_pwd,
                                            cert_reqs=cert_reqs, ca_certs=ca_cert,
                                            timeout=self._get_timeout(connect_timeout=self.__connect_timeout,
                                                                      read_timeout=self.__read_timeout),
                                            ssl_context=self.__get_ssl_context()
                                            )

could you please help on is the cert creation is the RC for the failure or should we look on other areas?

Thanks once again

**Comment by iambate** [comment]
So If I understand correctly, just using "-nomacver" when exporting/reading PKCS keystore is not an sufficient with openssl fips right? I am getting following error. Is this error related to using fips compliant openssl(i.e. not loading PKCS12KDF algo)?

```
Shrouded Keybag: pbeWithSHA1And3-KeyTripleDES-CBC, Iteration 1024
PKCS7 Encrypted data: pbeWithSHA1And3-KeyTripleDES-CBC, Iteration 1024
Error outputting keys and certificates
80117D8084700000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:373:Global default library context, Algorithm (PKCS12KDF : 0), Properties (<null>)
80117D8084700000:error:1180006B:PKCS12 routines:PKCS12_PBE_keyivgen_ex:key gen error:crypto/pkcs12/p12_crpt.c:55:
```

**Comment by paulidale** [comment]
Yes, that would be related.  3DES isn't permitted under FIPS anymore.

**Comment by WilburZjh** [comment]
I found this Document released at this year May -> https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4282.pdf.

It mentioned that PBKDF2 function is supported in FIPS mode for key derivation. Please correct me if Im wrong, the OpenSSL version of PBKDF2 is PKCS12KDF function?

**Comment by mattcaswell** [comment]
PBKDF2 is supported in our FIPS module already, but that is a different algorithm to PKCS12KDF. Both of those algorithms are supported in our default provider (see the KDF section below):

https://www.openssl.org/docs/man3.0/man7/OSSL_PROVIDER-default.html

But not in the FIPS provider (compare the list of KDFs supported):

https://www.openssl.org/docs/man3.0/man7/OSSL_PROVIDER-FIPS.html

**Comment by WilburZjh** [comment]
@mattcaswell thanks for the quick response, so if I want to run in FIPS mode, should I enable the FIPS provider and use PBKDF2 function from the FIPS module?

**Comment by mattcaswell** [comment]
The KDF used in PKCS12 is PKCS12KDF. You cannot configure it to use PBKDF2 instead.

**Comment by t8m** [comment]
More specifically PBKDF2 can also be used in PKCS12, but for a different purpose. The use of PKCS12KDF unfortunately cannot be replaced.

**Comment by ChauhanPriyanka** [comment]
hi @mattcaswell, @t8m ,

Currently we are migrating from OpenSSL v102 to v 3.0.x
We are extensively using PKCS12_create API, without FIPS it is working fine. The certificate generated by our application are consumed by Android, IOS and Windows device.

We are now making or application FIPS compliant, we cannot use the workaround of loading default provider in FIPS mode and we cannot skip integrity check also due to security issues. The application is audited and will not be certified in FIPS mode.

Could you please help what is the other alternate for us now.
- Will this draft https://datatracker.ietf.org/doc/draft-ietf-lamps-pkcs12-pbmac1/ fix the PKCS12_create API issue and it will start working in FPS mode.
- When is the fix  expected for this and in which version. As we are on 3.0.x version.

**Comment by mattcaswell** [comment]
> Will this draft https://datatracker.ietf.org/doc/draft-ietf-lamps-pkcs12-pbmac1/ fix the PKCS12_create API issue and it will start working in FPS mode.
When is the fix expected for this and in which version. As we are on 3.0.x version.

We believe it will fix the issue (on the assumption that PBMAC1 does not need to be inside the FIPS boundary - which the discussion above suggests should be ok). I'm not tracking the status of the draft but I understand it is still active.

@tomato42 is probably in a better position to know what the latest is on this draft. I assume it will be discussed at the IETF meeting this week.

> When is the fix expected for this and in which version. As we are on 3.0.x version.

Any fix will not be added to 3.0.x since that is a stable release and we don't add new features to stable releases. In order for the feature to be available, the draft first needs to be completed and published as a standard (OpenSSL only implements  crypto that has been standardised as a matter of policy). Again @tomato42 might be able to advise on the timelines here.

Once standardised a PR needs to be raised/approved and merged and it will then go into the next release. I feel fairly confident that @beldmit will raise a PR for this fairly soon after the standard is available. Even if everything goes smoothly and quickly, the earliest release that this could go into would be 3.3, which will be available in April 2024.

**Comment by beldmit** [comment]
I will raise a PR ASAP when the draft becomes an RFC. https://github.com/beldmit/openssl/tree/pkcs12_pbmac1 is a quick and dirty implementation

**Comment by ChauhanPriyanka** [comment]
hi @mattcaswell, @beldmit  
Thank you for the response.
Could you please clarify my understanding - even after the fix, which is related to PBMAC1.  PKCS12 will still use PKCS12KDF which is not FIPS complaint so I will still not able to use PKCS12_create API in FIPS mode.

**Comment by beldmit** [comment]
It's intended that you will be able to specify PBMAC1 instead of PKCS12KDF though I am not sure if it directly fits the PKCS12_create function or you will have to use smth else

**Comment by mattcaswell** [comment]
> that will be included in master branch, so 3.4.0 is earliest.

Current master branch is for 3.3.0 (to be release April 2024), so that would be the earliest.

**Comment by tomato42** [comment]
ah, right, 3.3.0 is earliest

**Comment by ChauhanPriyanka** [comment]
Hi @beldmit , 
It was my mistake I was using algo NID_pbe_WithSHA1And40BitRC2_CBC in PKCS12_create API hence it was not working in FIPS. with default algos. PKCS12_create api is working in FIPS. Apologies for the confusion.
Cert generated
```
MAC: sha256, Iteration 2048
MAC length: 32, salt length: 8
PKCS7 Encrypted data: PBES2, PBKDF2, AES-256-CBC, Iteration 2048, PRF hmacWithSHA256
Certificate bag
PKCS7 Data
Shrouded Keybag: PBES2, PBKDF2, AES-256-CBC, Iteration 2048, PRF hmacWithSHA256
```

Now, the next issue I encountered is in PKCS12_parse api, it is failing here. with the below mentioned error.
```
error:0308010C:digital envelope routines::unsupported
error:1180006B:PKCS12 routines::key gen error
error:1180006D:PKCS12 routines::mac generation error
error:11800071:PKCS12 routines::mac verify failure
```
So,  for my understanding, with the fix there will be a way to specify PBMAC1 instead of PKCS12KDF and mac integrity check will be performed successfully either during PKCS12_create or PKCS12_parse.

For now, could you please help me where should i set -nomac option in code (during create or during parse), to skip the mac verification.  This is just for testing purpose. We still can't skip mac verification, neither we can load default provider as this will get audited and our app will not be FIPS compliant.

Also, the fix will be available in 3.3, will it be eventually merged to 3.0.x version ?
We are working on getting FIPS certified so we need to this information to plan the release.

**Comment by paulidale** [comment]
PBMAC1 isn't available in the FIPS provider and I doubt it ever will be able to be validated, so that isn't a useful path for you.

We don't back port features to stable releases, so no a fix in 3.3 will not be merged to 3.0.x.

**Comment by ChauhanPriyanka** [comment]
Hi @paulidale,
Thank you for the response, could you please suggest what is the suitable alternate for this.

**Comment by mattcaswell** [comment]
> PBMAC1 isn't available in the FIPS provider and I doubt it ever will be able to be validated, so that isn't a useful path for you.

@paulidale - the argument earlier in this thread is that PBMAC1 is a trivial combination of a KDF and a MAC - and isn't a new algorithm in its own right (we wouldn't implement PBMAC1 as an EVP_MAC):

> Or to put it other way, PBMAC1 is the name of construct on PKCS#12 level, similar to PBES2, not a new MAC algorithm.

See [this comment](https://github.com/openssl/openssl/issues/19997#issuecomment-1373807358).

So the argument is that as long as the KDF and the MAC are inside the boundary then PBMAC1 can be outside the boundary and still be a FIPS compliant solution.

**Comment by paulidale** [comment]
That would work at least until the CMVP decides to draw this into the boundary too.
Then we'd have to validate it.

**Comment by tomato42** [comment]
@paulidale If that happens, then we will cross that bridge. As far as we (Red Hat) are aware, the existing fips module is already sufficient to implement PBMAC1 support in PKCS#12.

**Comment by ChauhanPriyanka** [comment]
PKCS12_create API also does not work in FIPS mode, following is the error:

`60690000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto\evp\evp_fetch.c:373:Global default library context, Algorithm (PKCS12KDF : 0), Properties (<null>)`

I have update my above comment https://github.com/openssl/openssl/issues/19997#issuecomment-1798015008

**Comment by t8m** [comment]
PKCS12KDF is the KDF that will get replaced once PBMAC1 is supported.

**Comment by ChauhanPriyanka** [comment]
Hi team,

Just for internal testing we are working on skipping MAC check. However this is not the valid option for us.

As suggested here, we set the mac_iter = -1, certificate is generated successfully with FIPS provider. 
p12 = PKCS12_create((char*)pswd, (char*)szCertStorePath, pkey, cert, NULL, 0, 0, 0, -1, 0).
https://www.openssl.org/docs/man3.0/man3/PKCS12_create.html#:~:text=mac_iter%20can%20be%20set%20to%20%2D1%20and%20the%20MAC%20will%20then%20be%20omitted%20entirely.%20This%20can%20be%20useful%20when%20running%20with%20the%20FIPS%20provider%20as%20the%20PKCS12KDF%20is%20not%20a%20FIPS%20approvable%20algorithm

While parsing the certificate we are facing this issue. What to set here to skip the MAC check.
**PKCS12_parse(ptr, password, out pkey, out cert, out cacerts)**

Error:
```
error:1180006C:PKCS12 routines::mac absent
error:11800071:PKCS12 routines::mac verify failure
```

**Comment by tomato42** [comment]
It's partially related to #20072 and in general will be fixed by implementing https://datatracker.ietf.org/doc/draft-ietf-lamps-pkcs12-pbmac1/

**Comment by Phani-Sreenivasa-Prasad** [comment]
> > 
> 
> @mattcaswell
> 
> Thanks for the information.
> 
> one doubt on this You can resolve this with appropriate global property settings. But ultimately loading some crypto from "default" means you are no longer FIPS.
> 
> * we are loading as of now fips and base providers and making the fips enable call from the code as shown below
> * self.crypto_context.EVP_default_properties_enable_fips(None, 1):
> 
> we understood that it is no longer fips compliant just for understanding a bit more, if we are using both providers when we are using the pkcs12kdf will it go to fips provider/default provider?



Did you get answer to this query? if so please share
we understood that it is no longer fips compliant just for understanding a bit more, if we are using both providers when we are using the pkcs12kdf will it go to fips provider/default provider?
